okhttp/src/main/java/okhttp3/Body.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package okhttp3;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.nio.charset.Charset;
+import okhttp3.internal.Util;
+import okhttp3.internal.http.HttpHeaders;
+import okio.Buffer;
+import okio.BufferedSink;
+import okio.BufferedSource;
+import okio.ByteString;
+import okio.Okio;
+
+import static okhttp3.internal.Util.UTF_8;
+
+/**
+ * The body for HTTP requests or responses.
+ */
+public class Body implements Closeable {
+  public static final byte[] EMPTY_ARRAY = new byte[0];
+  public static final Body EMPTY = new Body(null, EMPTY_ARRAY);
+
+  protected final MediaType contentType;
+  protected final long contentLength;
+  protected final BufferedSource readableContent;
+  protected final byte[] writableContent;
+
+  private Reader reader;
+
+  public Body() {
+    this(null, -1, null, null);
+  }
+
+  public Body(MediaType contentType) {
+    this(contentType, -1, null, null);
+  }
+
+  public Body(MediaType contentType, byte[] content) {
+    this(contentType, content.length, content, new Buffer().write(content));
+  }
+
+  public Body(MediaType contentType, Buffer content) {
+    this(contentType, content.size(), content.readByteArray(), content);
+  }
+
+  public Body(MediaType contentType, BufferedSource content) {
+    this(contentType, -1, null, content);
+  }
+
+  public Body(MediaType contentType, long contentLength, BufferedSource content) {
+    this(contentType, contentLength, null, content);
+  }
+
+  public Body(MediaType contentType, long contentLength, byte[] writableContent,
```

Comment:
Way too many constructors _and_ static factories. I liked the old setup where static factories mean we handle everything for you from known types and subclasses opt-in to handling everything that's required.